### PR TITLE
RPA Paycodes

### PR DESCRIPTION
**Pull request by fyookball**
This implements RPA Re-usable payment address (AKA Paycodes) with a CLI.

Note this only implements the basic functions of generating a paycode, a transaction, and extracting private key.  Other functions such as storing the key in the keystore will be implemented in the next stages.

### CODE REVIEW

File: lib/cashaddr.py [hunk 1]
```diff
@@ -84,6 +84,7 @@ def _pack_addr_data(kind, addr_hash):
         encoded_size |= 0x04
     encoded_size |= (len(addr_hash) - 20 * offset) // (4 * offset)
 
+
```
Comment (cculianu) [inline lib/cashaddr.py:87]: Why add the extra space? Please clean this up.

File: lib/cashaddr.py [hunk 2]
```diff
@@ -143,7 +144,7 @@ def _decode_payload(addr):
 PUBKEY_TYPE = 0
 SCRIPT_TYPE = 1
 
-def decode(address):
+def decode(address,rpa_paycode = False):
```
Comment (cculianu) [inline lib/cashaddr.py:147]: Why touch this?

Comment (cculianu) [reply lib/cashaddr.py:147]: Also please put a space after the comma `,`.

File: lib/cashaddr.py [hunk 3]
```diff
@@ -118,7 +119,7 @@ def _decode_payload(addr):
         raise ValueError('address prefix is missing: {}'.format(addr))
     if not all(33 <= ord(x) <= 126 for x in prefix):
         raise ValueError('invalid address prefix: {}'.format(prefix))
-    if not (8 <= len(payload) <= 124):
+    if not (8 <= len(payload) <= 256):
```
Comment (cculianu) [inline lib/cashaddr.py:122]: ??  CashAddr payloads have a well specified length.  Why was this file modified at all?

File: lib/commands.py [hunk 4]
```diff
@@ -540,23 +541,42 @@ def _mktx(self, outputs, fee=None, change_addr=None, domain=None, nocheck=False,
         if not unsigned:
             run_hook('sign_tx', self.wallet, tx)
             self.wallet.sign_transaction(tx, password)
+            if addtransaction:
```
Comment (cculianu) [inline lib/commands.py:544]: Why does this commit contain cherry-picks from master.  Please learn to use the `git rebase` command.  This needs a rebase.

File: lib/commands.py [hunk 5]
```diff
@@ -986,6 +1008,7 @@ def add_global_options(parser):
     group.add_argument("-wp", "--walletpassword", dest="wallet_password", default=None, help="Supply wallet password")
     group.add_argument("--testnet", action="store_true", dest="testnet", default=False, help="Use Testnet")
     group.add_argument("--testnet4", action="store_true", dest="testnet4", default=False, help="Use Testnet4")
+    group.add_argument("--scalenet", action="store_true", dest="scalenet", default=False, help="Use Scalenet")
```
Comment (cculianu) [inline lib/commands.py:1011]: Please rebase.

File: lib/schnorr.py [hunk 6]
```diff
@@ -118,7 +118,7 @@ def nonce_function_rfc6979(order, privkeybytes, msg32, algo16=b'', ndata=b''):
     return k
 
 
-def sign(privkey, message_hash):
+def sign(privkey, message_hash,ndata=b''):
```
Comment (cculianu) [inline lib/schnorr.py:121]: Space after comma

File: lib/transaction.py [hunk 7]
```diff
@@ -511,6 +511,36 @@ def is_schnorr_signed(self, input_idx):
                        for sig in self._inputs[input_idx].get('signatures', []))
         return False
 
+    def rpa_paycode_swap_dummy_for_destination(self,rpa_dummy_address, rpa_destination_address):
```
Comment (cculianu) [inline lib/transaction.py:514]: Space after comma

File: lib/transaction.py [hunk 8]
```diff
@@ -511,6 +511,36 @@ def is_schnorr_signed(self, input_idx):
...
+        # It swaps out a dummy destination address for the rpa-generated address.
+        # WARNING: This is not recommended for use outside of RPA since it could
+        # be dangerous to change the destination address of a transaction.
+ 
+        # This method expects cashaddr strings for parameters rpa_dummy_address and rpa_destination_address
+  
+        length = len (self._outputs)
+        i = 0
+        
+        while i < length:
```
Comment (cculianu) [inline lib/transaction.py:525]: Python idiomatic programming: Use `for i in range(length)`

File: lib/transaction.py [hunk 9]
```diff
@@ -511,6 +511,36 @@ def is_schnorr_signed(self, input_idx):
...
+        length = len (self._outputs)
+        i = 0
+        
+        while i < length:           
+            # Grab the address of each output...
+            loop_iteration_output = self._outputs[i]
+            loop_iteration_address = loop_iteration_output[1]
+             
+            # Compare the address to see if its the one we need to swap.  Note: 0 for the to_string = CASHADDR format
+            if loop_iteration_address.to_string(0) == rpa_dummy_address:
```
Comment (cculianu) [inline lib/transaction.py:531]: Please don't use magic int literals.  `to_string(Address.FMT_CASHADDR)` for example.

File: lib/transaction.py [hunk 10]
```diff
@@ -909,14 +939,14 @@ def _ecdsa_sign(sec, pre_hash):
         return sig
 
     @staticmethod
-    def _schnorr_sign(pubkey, sec, pre_hash):
+    def _schnorr_sign(pubkey, sec, pre_hash,ndata=b''):
```
Comment (cculianu) [inline lib/transaction.py:942]: Space after comma

File: lib/transaction.py [hunk 11]
```diff
@@ -511,6 +511,36 @@ def is_schnorr_signed(self, input_idx):
...
+                rpa_replacement_output = list(loop_iteration_output)
+                rpa_replacement_output[1]= rpa_replacement_address
+                self._outputs[i]=tuple(rpa_replacement_output)
+            i +=1
+            
+        # It is necessary to re-initialize "raw" so the output swap becomes part of transaction when it is later serialized
+        self.raw = None
+        return 0
+        
+
```
Comment (cculianu) [inline lib/transaction.py:543]: Remove extra newline

File: lib/wallet.py [hunk 12]
```diff
@@ -2208,7 +2216,7 @@ def add_hw_info(self, tx):
                 info[addr] = index, sorted_xpubs, self.m if isinstance(self, Multisig_Wallet) else None, self.txin_type
         tx.output_info = info
 
-    def sign_transaction(self, tx, password, *, use_cache=False):
+    def sign_transaction(self, tx, password, *, use_cache=False, ndata=b''):
```
Comment (cculianu) [inline lib/wallet.py:2219]: I'm surprised this has a space after the comma!

File: lib/wallet.py [hunk 13]
```diff
@@ -2218,6 +2226,7 @@ def sign_transaction(self, tx, password, *, use_cache=False):
 
         Warning: If you modify non-signature parts of the transaction
         afterwards, do not use `use_cache`! """
+
```
Comment (cculianu) [inline lib/wallet.py:2229]: Please don't add extra whitespace randomly ...

File: lib/wallet.py [hunk 14]
```diff
@@ -2230,7 +2239,7 @@ def sign_transaction(self, tx, password, *, use_cache=False):
         for k in self.get_keystores():
             try:
                 if k.can_sign(tx):
-                    k.sign_transaction(tx, password, use_cache=use_cache)
+                    k.sign_transaction(tx, password, use_cache=use_cache,ndata=ndata)
```
Comment (cculianu) [inline lib/wallet.py:2242]: Space after comma

File: lib/commands.py [hunk 15]
```diff
@@ -540,23 +541,42 @@ def _mktx(self, outputs, fee=None, change_addr=None, domain=None, nocheck=False,
...
+    @command('')
+    def rpa_generate_paycode(self):
+        return rpa_paycode.rpa_generate_paycode(self.wallet)
+
     @command('wp')
-    def payto(self, destination, amount, fee=None, from_addr=None, change_addr=None, nocheck=False, unsigned=False, password=None, locktime=None,
+    def rpa_generate_transaction_from_paycode(self, amount,  paycode=None, fee= None, from_addr=None, change_addr=None, nocheck=False, unsigned=False, password=None, locktime=None,
               op_return=None, op_return_raw=None):
+        # WARNING: Amount is in full Bitcoin Cash units
+        return rpa_paycode.rpa_generate_transaction_from_paycode(self.wallet, self.config, amount,paycode)
```
Comment (cculianu) [inline lib/commands.py:558]: space-after-comma

File: lib/commands.py [hunk 16]
```diff
@@ -540,23 +541,42 @@ def _mktx(self, outputs, fee=None, change_addr=None, domain=None, nocheck=False,
...
+            
+    @command('wp')
+    def rpa_extract_private_key_from_transaction(self,raw_tx, password=None):     
+        return rpa_paycode.rpa_extract_private_key_from_transaction(self.wallet,raw_tx,password)
```
Comment (cculianu) [inline lib/commands.py:562]: space-after-comma

File: lib/rpa_paycode.py [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+    coins = wallet.get_spendable_coins(domain, config)
+    tx = wallet.make_unsigned_transaction(coins, final_outputs, config, fee, change_addr)
+    if locktime != None:
+        tx.locktime = locktime
+    if not unsigned:
+        run_hook('sign_tx', wallet, tx) 
+        wallet.sign_transaction(tx, password)
+    return tx
+
+def calculate_paycode_shared_secret(private_key,public_key,outpoint):
```
Comment (cculianu) [inline lib/rpa_paycode.py:65]: space-after-comma.. all over this file

File: lib/schnorr.py [hunk 18]
```diff
@@ -139,7 +139,7 @@ def sign(privkey, message_hash):
     if _secp256k1_schnorr_sign:
         sig = create_string_buffer(64)
         res = _secp256k1_schnorr_sign(
-            secp256k1.secp256k1.ctx, sig, message_hash, privkey, None, None
```
Comment (cculianu) [inline lib/schnorr.py:142]: Is the default argument of `b''` equivalent to the previous `None` here?

File: lib/schnorr.py [hunk 19]
```diff
@@ -161,7 +161,7 @@ def sign(privkey, message_hash):
         pubbytes = point_to_ser(pubpoint, comp=True)
 
         k = nonce_function_rfc6979(order, privkey, message_hash,
-                                   algo16=b'Schnorr+SHA256\x20\x20')
+                                   algo16=b'Schnorr+SHA256\x20\x20',ndata=ndata)
```
Comment (cculianu) [inline lib/schnorr.py:164]: Is the addition of the default `ndata=b''` equivalent to the previous call?

File: lib/transaction.py [hunk 20]
```diff
@@ -511,6 +511,36 @@ def is_schnorr_signed(self, input_idx):
...
+            # Grab the address of each output...
+            loop_iteration_output = self._outputs[i]
+            loop_iteration_address = loop_iteration_output[1]
+             
+            # Compare the address to see if its the one we need to swap.  Note: 0 for the to_string = CASHADDR format
+            if loop_iteration_address.to_string(0) == rpa_dummy_address:
+                # Do the swap
+                rpa_replacement_address = Address.from_string(rpa_destination_address)
+                rpa_replacement_output = list(loop_iteration_output)
+                rpa_replacement_output[1]= rpa_replacement_address
```
Comment (cculianu) [inline lib/transaction.py:535]: Please properly put spaces around operators -- that's the standard in Python e.g. `output[1] = rpa` not `output[1]= rpa`

That goes for everything in this file.

File: lib/cashaddr.py (hunk 3 above)
Comment (cculianu) [reply lib/cashaddr.py:122]: If you must do this pass a flag own to use a different value for your special case.  I would prefer if you didn't have to touch this file.. this is bizarre to me that you would have to.... Why?

Comment (fyookball) [reply lib/cashaddr.py:122]: The cashaddr code doesn't seem to work out of the box for "addresses" as long as the paycodes.  There were two problems encountered, both of which appear to be sanity check related.  The one here (124) was bumped from the 80 value of the original bech32.   here->https://github.com/Electron-Cash/Electron-Cash/commit/ae60cc9386f1082421a3bddb41a1baa94ce4edcd#diff-6624c8abadfc8d6860e0879af63e447b

The other sanity check is related to length but causes paycode to fail and doesn't seem necessary.  But this is why I created a flag (rpa_paycode = False).  

I can make them both dependent on a flag or just separate out the functions.  Either way(s) seem ok to me.

File: lib/commands.py (hunk 4 above)
Comment (fyookball) [reply lib/commands.py:544]: Ok.

File: lib/schnorr.py (hunk 19 above)
Comment (fyookball) [reply lib/schnorr.py:164]: Yes, see https://github.com/Electron-Cash/Electron-Cash/blob/master/lib/schnorr.py#L99  

If there is nothing present, it just becomes empty (byte) string

File: lib/schnorr.py (hunk 18 above)
Comment (fyookball) [reply lib/schnorr.py:142]: Seems so from looking at libsec256k1 but I will retest to make sure the values don't change.

File: lib/transaction.py (hunk 20 above)
Comment (fyookball) [reply lib/transaction.py:535]: Thanks for the review! will do

File: lib/schnorr.py (hunk 18 above)
Comment (cculianu) [reply lib/schnorr.py:142]: I'm just worried about unintended side effects.  I believe `None` ends up in secp as a C `NULL` where as `b''` ends up in C as .. a string of size 0 e.g. `""`.

Why change the default behavior of this function?  Can't you have the `ndata` argument default to `None`?  And then for paycode it will work as you like and for secp it will work as you like.

File: lib/cashaddr.py (hunk 3 above)
Comment (cculianu) [reply lib/cashaddr.py:122]: Again, why change the default execution path for your custom use-case?

There will be unintended side effects.  This change now impacts the app globally --

You already have that flag you are passing .. please pass it down and default the normal execution path without the flag set to the old behavior in all cases in this file, minimally.

Comment (cculianu) [reply lib/cashaddr.py:122]: Also -- how did you arrive at a 256 value?  Is that a magical random number you chose because it's a power of two and "looks right"?

Can you please motivate this further? (That is: talk about it and explain).

File: lib/cashaddr.py [hunk 21]
```diff
@@ -166,10 +167,11 @@ def decode(address):
     version = decoded[0]
     addr_hash = bytes(decoded[1:])
     size = (version & 0x03) * 4 + 20
+    
     # Double the size, if the 3rd bit is on.
     if version & 0x04:
         size <<= 1
-    if size != len(addr_hash):
+    if size != len(addr_hash) and not rpa_paycode:
```
Comment (cculianu) [inline lib/cashaddr.py:174]: Again this makes me very suspicious -- this is another thing that is well specified by CashAddr.  Why avoid this check in the rpa case?

Can you please point me to the rpa spec?  Can you give me an example of an RPA address so I may try it against the regular bitcoind "Decode" function this is analogous to?

Comment (cculianu) [reply lib/cashaddr.py:174]: I think the way you are supposed to do this is you are supposed to be encoding a version properly into the payload to make the length check succeed here.  

This seems to me like the RPA address spec needs revision if it's producing payloads that don't decode properly in the cashaddr scheme....

File: lib/schnorr.py (hunk 18 above)
Comment (cculianu) [reply lib/schnorr.py:142]: No. This is not equivalent.  I just verified the C side.  It's as I suspected. `None` here ends up as -> `NULL` in C.  Whether or not that pointer is `NULL` or `""` on the C side completely changes the behavior of the nonce function.

See this bit of code which is where the behavior is affected: https://github.com/Bitcoin-ABC/secp256k1/blob/69ccf0d2f758b1a4df9cde8222459381e08fc1dc/src/secp256k1.c#L330

That `ndata` ends up exactly as the `data` pointer in the C.  If the `data` pointer is not NULL, then **unconditionally** 32 bytes of data are appended (but that pointer points to 0 bytes!).

You actually should be enforcing that `len(ndata)` == 32, or that `ndata` is None..  all over this codebase as well.  This will lead to crashes.

**Review by cculianu — CHANGES_REQUESTED** [review]
See my review comment here regarding interfacing with secp. The way this is coded it will lead to possible segmentation faults and incorrect behavior.

https://github.com/Electron-Cash/Electron-Cash/pull/2001/files#r494084058

File: lib/schnorr.py (hunk 6 above)
Comment (cculianu) [inline lib/schnorr.py:121]: `ndata` needs to default to `None`.  If it is not `None`, it needs to be exactly 32 bytes, otherwise an exception should be thrown.  This invariant needs to be enforced all over the place where you added this `ndata` argument.

File: lib/schnorr.py (hunk 19 above)
Comment (cculianu) [reply lib/schnorr.py:164]: Right, in that execution path in the python that's ok -- but on the secp side it is not the same thing.

File: lib/cashaddr.py (hunk 3 above)
Comment (fyookball) [reply lib/cashaddr.py:122]: My thought was that if this is a kind of arbitrary value, then maybe it's better to just increase it rather than start adding custom rules...and the 256 was arbitrary as well.  If you prefer just to add a custom flag and pass it down, can do!

Comment (cculianu) [reply lib/cashaddr.py:122]: How do you know that it's arbitrary, though?  It seems deliberately chosen.  Do you have anything I can look at to see that?

Comment (fyookball) [reply lib/cashaddr.py:122]: I don't know for sure.  I had asked Shammah about that commit and his reply seemed to indicate that.

In any case, I will implement a flag so the default behavior is not impacted.  I will also investigate "encoding a version properly into the payload to make the length check succeed"  as you recommended.
